### PR TITLE
SI-8153 Mutation is hard, let's go shopping with an empty list

### PR DESCRIPTION
**Pull request by Ichoran**
Changed the implementation of iterator to be more robust to mutation of the underlying ListBuffer.

Test added to verify the fix.

### CODE REVIEW

**Comment by adriaanm** [comment]
Ping @Ichoran

**Comment by Ichoran** [comment]
@adriaanm - Yeah, I know there are issues.  Not sure why that tests fails, but it seems to be real.

File: src/library/scala/collection/mutable/ListBuffer.scala [hunk 1]
```diff
@@ -389,22 +389,15 @@ final class ListBuffer[A]
     // a structure while iterating, but we should never return hasNext == true
     // on exhausted iterators (thus creating exceptions) merely because
     // values were changed in-place.
-    var cursor: List[A] = null
-    var delivered = 0
-
-    // Note: arguably this should not be a "dynamic test" against
```
Comment (adriaanm) [inline src/library/scala/collection/mutable/ListBuffer.scala:395]: @Ichoran, the hint here corresponds to the actual test failures predicted by said comment at def units in Global.scala:

```
    /* An iterator returning all the units being compiled in this run */
    /* !!! Note: changing this to unitbuf.toList.iterator breaks a bunch
       of tests in tests/res.  This is bad, it means the resident compiler
       relies on an iterator of a mutable data structure reflecting changes
       made to the underlying structure (in whatever accidental way it is
       currently depending upon.)
     */
    def units: Iterator[CompilationUnit] = unitbuf.iterator
```

Comment (Ichoran) [inline src/library/scala/collection/mutable/ListBuffer.scala:395]: Thanks...have been too busy to look into it.  Will do so today or tomorrow at latest..

Comment (adriaanm) [inline src/library/scala/collection/mutable/ListBuffer.scala:395]: Ok!

**Comment by Ichoran** [comment]
@adriaanm - I fixed the issue not by changing ListBuffer (there I just documented what the new behavior is), but by fixing the compiler to make its assumptions more explicit.  Crossing fingers that this actually works.  (And that adding an ArrayBuffer dependency doesn't mess up load times or anything.)

**Comment by Ichoran** [comment]
PLS REBUILD/pr-scala@b8881a2

**Comment by Ichoran** [comment]
Trying squashing to a single commit to see if that will resolve the errors that I can't reproduce locally.  (Probably not.)

**Comment by retronym** [comment]
PR validation tests your changes merged onto master. Try rebasing to make sure you're testing exactly the same thing locally.

**Comment by Ichoran** [comment]
@retronym - Thanks for the advice, but I am always up-to-date with master.  What I need, if I can't reproduce the errors locally, is a way to reliably get text output back from the jenkins job.  Something like

  http://jenkins-ci.361315.n4.nabble.com/Display-custom-logs-on-Jenkins-td3771103.html

but I don't think I have the permissions necessary to set up something like that.

**Comment by adriaanm** [comment]
Which logs are you looking for? https://scala-webapps.epfl.ch/jenkins/job/pr-scala-publish-core/2075/console?

**Comment by Ichoran** [comment]
@adriaanm - If everything I put into standard error will show up there, that'd be adequate.  It seems abbreviated to me, but maybe I'm wrong.

**Comment by adriaanm** [comment]
It captures both stdout and stderr. (As far as I know.)

**Comment by Ichoran** [comment]
@adriaanm - Not for tests, but the testing frameworks are what's eating that output.  Anyway, I've managed to reproduce the problems on my workstation (not laptop).

**Comment by adriaanm** [comment]
ah, the junit tests? they are in log files, which are artifacts of the jenkins builds: https://scala-webapps.epfl.ch/jenkins/job/pr-scala-test/5288/artifact/scala/build/junit/

**Comment by Ichoran** [comment]
@adriaanm - I meant the other tests, which end up as -dir.log files under test/files/dir (for dir = run, pos, neg, res, etc.).  There aren't enough JUnit tests for them to fail on much.

**Comment by adriaanm** [comment]
Ah good point, we should tell jenkins to archive these as well

**Comment by adriaanm** [comment]
LGTM! Thanks for persevering in the face of jenkins.

**Pull request merged**